polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -107,6 +201,77 @@ struct RaiseKernelName : public OpRewritePattern<LLVM::AddressOfOp> {
                                        : std::nullopt;
   }
 };
+
+/// Raise constructs assigning a kernel name to a handler to
+/// `sycl.host.handler.set_kernel`.
+///
+/// This pattern acts on `FunctionOpInterface` instances as it will not be
+/// removing the operations assigning the kernel name, but creating additional
+/// ones to mark the construct.
+struct RaiseSetKernel : public OpInterfaceRewritePattern<FunctionOpInterface> {
+public:
+  using OpInterfaceRewritePattern<
+      FunctionOpInterface>::OpInterfaceRewritePattern;
+
+  LogicalResult matchAndRewrite(FunctionOpInterface op,
+                                PatternRewriter &rewriter) const final {
+    bool changed = false;
+    // Go through each `sycl.host.get_kernel` introduced
+    op.walk([&](sycl::SYCLHostGetKernelOp getKernel) {
+      SymbolRefAttr symbolRef = getKernel.getKernelNameAttr();
+      // `llvm.invoke` operations using that are candidates for this pattern
+      for (auto invoke : getUsersOfType<LLVM::InvokeOp>(getKernel)) {
+        // We mark already visited `llvm.invoke` operations to avoid infinite
```

Comment:
Wouldn't it be more idiomatic for MLIR pattern rewrites to have the pattern match `llvm.invoke` and return `failure()` for all the invokes we're not interested in instead of using custom logic here?

Reply:
I wish... According to [the docs](https://mlir.llvm.org/docs/PatternRewriter/#restrictions):
```quote
The root operation is required to either be: updated in-place, replaced, or erased.
```